ini_file - add feature 'section_has'

jakob-bebop opened the pull request:
##### SUMMARY
ini_file: Add an optional parameter `section_has`. If the target ini file contains more than one `section`, use `section_has` to specify which one should be updated.

Use case: In wireguard config there can be multiple [Peer] sections. See example in source code doc string

##### ISSUE TYPE
- Feature Pull Request

##### COMPONENT NAME
ini_file

felixfontein commented on plugins/modules/ini_file.py:
```diff
@@ -46,6 +46,24 @@
+      - Among possibly multiple sections of the same name, select the one that contains these values.
+      - With O(state=present), if a suitable section is not found, a new section will be added, including the required options.
+      - With O(state=absent), at most one O(section) is removed if it contains the values.
+    version_added: 8.0.0
```
8.0.0 has already been released; the next release accepting new features will be 8.1.0:
```suggestion
    version_added: 8.1.0
```

felixfontein commented on plugins/modules/ini_file.py:
```diff
@@ -302,14 +354,22 @@ def do_ini(module, filename, section=None, option=None, values=None,
     for index, line in enumerate(ini_lines):
+        # end of section:
+        if within_section and line.startswith(u'['):
+            if check_section_has(
```
This also needs to be run when the `for` loop doesn't end with a `break`, otherwise you'll miss the section if it is the very last section of the file.

(This might already be a problem with the existing code, when quickly glancing at this diff.)

jakob-bebop replied:
I think the sentinel value added in line 335 makes it so that the section always ends here. Let me look into adding an integration test to verify this.

felixfontein replied:
Ah right, I missed that one. Thanks for pointing this out!

felixfontein commented on plugins/modules/ini_file.py:
```diff
@@ -430,6 +490,9 @@ def do_ini(module, filename, section=None, option=None, values=None,
     if not within_section and state == 'present':
         ini_lines.append(u'[%s]\n' % section)
         msg = 'section and option added'
+        if 'section_has' in module.params:
```
This check is always true, since `module.params['section_has']` is `None` when the option has not been specified.

You want to check for `module.params['section_has'] is not None`.

jakob-bebop replied:
Yes!

felixfontein reviewed (commented):
I didn't had time to read the change in detail, but here are some first comments from a quick glance:

jpmens commented on plugins/modules/ini_file.py:
```diff
@@ -179,6 +197,28 @@
     mode: '0600'
     state: present
+
+- name: remove the peer configuration for 10.128.0.11/32
```
is the CIDR in this comment correct? I would have thought it ought to be `10.4.0.11/32`

jakob-bebop replied:
corrected

jpmens commented on plugins/modules/ini_file.py:
```diff
@@ -46,6 +46,24 @@
+        description: Locate the section with this specific value.
+        required: true
+    description:
+      - Among possibly multiple sections of the same name, select the one that contains these values.
```
I feel`section_has` would be easier to understand if its description included something along the lines of "section has these values"

felixfontein replied:
Maybe it would also be a good idea to rename the option to `section_has_values`.

jakob-bebop replied:
How do you feel about this idea? I have been using it to set up some Wireguard peers (similar to the examples), and figured it might be a useful addition.

I guess a lot of little adjustments could be made... There's also part of the behaviour wrt "section" that's not terribly clear from the documentation (specifically, if you set "section", leave out "option", and set "state=absent", the entire section is removed from the file). 

I'd also prefer to add a couple of test cases.

Do you think it's worth the time fiddling with that? ... And if yes, is it correct that the place to add test cases would be in `tests/integration/targets/ini_file/tasks/tests`?

Wrt the name of the option ( /section_has/ or /section_has_values/ ), I thought the former to be easy to understand when I wrote my playbook task (which is similar to the example). It's quite clear from the context that it's related to an option and value. And it's shorter, which I like.

felixfontein replied:
IMO adding this is OK. Not sure what others think though :)

> I'd also prefer to add a couple of test cases, including some for the existing *section* functionality

:+1: for that!

> And if yes, is it correct that the place to add test cases would be in `tests/integration/targets/ini_file/tasks/tests`?

Yes, that's the right place. If you add a new file there, also remember to add it to tests/integration/targets/ini_file/tasks/main.yml.

(You can run the tests with `ansible-test integration -v ini_file`, assuming you checked out the repository in a directory tree structure that ansible-test needs. ansible-test will tell you if you didn't.)

> Wrt the name of the option (*section_has* or *section_has_values*), I thought the former to be easy to understand when I wrote my playbook task (which is similar to the example). It seems clear from the context, that it's related to an option and value. And it's shorter, which I like.

In the playbook itself `section_has` is sufficient, but just looking at the name without seeing an example, I think the longer name is better. Since the option name shows up in both situations (docs vs. examples/playbooks), I think it is better to use the slightly longer one.

Would be nice to also hear more opinions here, such questions are highly subjective :)

felixfontein replied:
@jpmens any opinions on this?

jpmens replied:
Yes, I too prefer the longer name, and that in spite of me typically liking short variable names, etc. :) As already mentioned I think the longer parameter name explains itself better. I thus agree with @felixfontein

felixfontein commented on plugins/modules/ini_file.py:
```diff
@@ -430,6 +490,9 @@ def do_ini(module, filename, section=None, option=None, values=None,
     if not within_section and state == 'present':
         ini_lines.append(u'[%s]\n' % section)
         msg = 'section and option added'
+        if 'section_has' in module.params and module.params['section_has']:
```
```suggestion
        if module.params['section_has']:
```

felixfontein commented:
needs_info

jakob-bebop commented:
- changed the name of the config option to use the more explicit name proposed above.
- added some test cases in `tests/integration/targets/ini_file/tasks/tests/07-section.yml`. The existing _section_ functionality seems to be covered in the `02-values.yml`
- I set _version_added_ to 8.5 in the docstring

felixfontein commented on plugins/modules/ini_file.py:
```diff
@@ -176,12 +194,56 @@
     mode: '0600'
     state: present
 
+- name: remove the peer configuration for 10.128.0.11/32
```
```suggestion
- name: Remove the peer configuration for 10.128.0.11/32
```

felixfontein commented on plugins/modules/ini_file.py:
```diff
@@ -176,12 +194,56 @@
     value: lemon juice
     state: present
+
+- name: remove the peer configuration for 10.128.0.11/32
```
```suggestion
- name: Remove the peer configuration for 10.128.0.11/32
```

felixfontein commented on plugins/modules/ini_file.py:
```diff
@@ -176,12 +194,56 @@
+  community.general.ini_file:
+    path: /etc/wireguard/wg0.conf
+    section: Peer
+    section_has:
```
```suggestion
    section_has_values:
```
Same in the examples below.

utoddl commented:
Late to the party, and I know this is a matter of personal taste, but `section_has_values` is not (in my opinion) a great name. Would it be too late to consider something else? Perhaps `section_discriminator`?

Considering especially the case where `allow_no_value: true`, then `section_has_values` is particularly ill-suited. The function of the list is to discriminate one particular section from other identically named sections. So name it for its function.

utoddl commented:
There's a problem with this feature as implemented. Both `option` and `value` of the `section_has_values` list of dicts are marked `required=True`. But some options don't require a value, as per when `allow_no_value: true` is specified. The values in the list of dicts should be optional in that case.

felixfontein commented:
I don't think having `value` required is a *problem* - it's just not as flexible as it could be. But such flexibility can be added later on (in another PR).

The main problem with later extensions is the name, as you point out - `section_has_values` is about having specific values. `section_discriminator` or `section_selector` or something like that is more general and will also work if we allow to select sections by keys without values, or by keys (without caring for values), or by absent keys, or ...

jakob-bebop commented:
> There's a problem with this feature as implemented. Both `option` and `value` of the `section_has_values` list of dicts are marked `required=True`. But some options don't require a value, as per when `allow_no_value: true` is specified. The values in the list of dicts should be optional in that case.

Options with "no value" is not supported by this feature. I don't see why that is a problem---such options not part of the use case I had in mind when implementing the feature.

jakob-bebop commented:
Consider the behaviour of the existing `section` functionality, if the section is not found: A new section of that name is added to the file, and the option (value or no value) is added to that section. 

With the proposed feature enabled, the additional required values are also added to the new section. In the YAML file it looks like

```
    - community.general.ini_file:
        section: Peer
        section_has_values:
          - option: AllowedIps
            value: '{{ my_ip_var }}/32'
        option: PublicKey
        value: '{{ my_pubkey.stdout }}'
        dest: '{{ wg_conf }}'
      notify:
        - restart wg
```

meaning "the file should have a _Peer_ section with the provided values for _AllowedIps_ and _PublicKey_, and the section (with that _AllowedIps_ value) should be updated if it already exists.

I find the above snippet reasonably easy to read, and I think that including "selection" og "discrimination" in the option name would make it less obvious what happens in the "section not found" case. 

Perhaps `section_requires` ? Personally, I still prefer the initial proposition of simply `section_has`.

russoz commented:
Even later than others to the party, but a suggestion of a compromise:

```yaml
    - community.general.ini_file:
        section: Peer
        select_section_with:
          values:
            - option: AllowedIps
              value: '{{ my_ip_var }}/32'
        option: PublicKey
        value: '{{ my_pubkey.stdout }}'
        dest: '{{ wg_conf }}'
      notify:
        - restart wg
```
In that case, `select_section_with` becomes a `dict` with `values` in `suboptions`. The name is longer but, IMHO, better conveys the concept, while at the same time allowing for future expansions. 

What do you think, @utoddl , @jakob-bebop ?

russoz commented on plugins/modules/ini_file.py:
```diff
@@ -305,14 +379,22 @@ def do_ini(module, filename, section=None, option=None, values=None,
 
     for index, line in enumerate(ini_lines):
+        # end of section:
+        if within_section and line.startswith(u'['):
```
All code in Ansible is unicode by definition. This is redundant:
```suggestion
        if within_section and line.startswith('['):
```
Same applies to other occurrences - yes, I see the existing code makes use of that. We should review all occurrences in another PR.

felixfontein replied:
Actually we should NOT do that while we still support Python 2.7.

russoz replied:
In my memory, `unicode` was in the mandatory imports from `__future__`. That clearly not being the case, I take it back.

russoz reviewed (commented):
we should stop using `u'...'`

utoddl commented:
There are some funky undefined interactions between this proposal and the `exclusive` option, too.
I put together a PR - https://github.com/jakob-bebop/ansible.community.general/pull/1 - to address `exclusive`, `allow_no_value`, and the situation where the `section_has_value` specification is overridden by the `option` and `value` specs. (As of this writing, I believe the currently proposed code will inadvertently create multiple instances of a particular key with the same or different values. Oops!)

My proposed upstream (to @jakob-bebop) PR also adopts the technique of accepting either the single string "`value`" or the list "`values`" in each `section_has_value` list element in the same way that `ini_file` itself will accept either `value` or `values` parameters. In addition, it makes the `value[s]` in the `section_has_value` list optional, so you can match purely on the existence of one or more keys in a section regardless of their corresponding values (which may or may not exist, b/c `allow_no_value`).

The introduction of `section_has_values` in combination with `exclusive` and `allow_no_value` creates opportunity for quite a bit of expressive power for what turns out to be surprisingly readable code! I'm taking it as an indicator that it's probably correct in that it does what the user would expect — at least in the corner cases I've been able to come up with.

Aside: The other surprising thing to me, considering that I don't often go prying into upstream work, is that [my own single active pull request](https://github.com/ansible-collections/community.general/pull/8075) at the moment _conflicts with these changes in this same file_! What are the chances? Still, I think this `section_has_values` work is a worthwhile contribution, and I'll happily rework my own PR - or the other way around - depending on which one gets merged first.

felixfontein commented:
@jakob-bebop please take a look at https://github.com/jakob-bebop/ansible.community.general/pull/1. And also please note that there is now a conflict with the current `main` branch.

felixfontein commented on plugins/modules/ini_file.py:
```diff
@@ -44,6 +44,30 @@
+      - Among possibly multiple sections of the same name, select the first one that contains matching options and values.
+      - With O(state=present), if a suitable section is not found, a new section will be added, including the required options.
+      - With O(state=absent), at most one O(section) is removed if it contains the values.
+    version_added: 8.5.0
```
```suggestion
    version_added: 8.6.0
```

felixfontein commented on plugins/modules/ini_file.py:
```diff
@@ -44,6 +44,30 @@
+        required: true
+      value:
+        type: str
+        description: Matching O(option) must have this specific value.
```
This is referring to `option` in `section_has_values[]`, not to the top-level `option`:
```suggestion
        description: Matching O(section_has_values[].option) must have this specific value.
```

felixfontein commented on plugins/modules/ini_file.py:
```diff
@@ -44,6 +44,30 @@
+        description: Matching O(option) must have this specific value.
+      values:
+        description:
+          - The string value to be associated with an O(option).
+          - Mutually exclusive with O(value).
+          - O(value=v) is equivalent to O(values=[v]).
```
```suggestion
          - The string value to be associated with an O(section_has_values[].option).
          - Mutually exclusive with O(section_has_values[].value).
          - O(section_has_values[].value=v) is equivalent to O(section_has_values[].values=[v]).
```

felixfontein reviewed (commented):
Just some more nits. Please note that there are some other suggestions that are still open.

felixfontein commented:
If nobody objects, I'll merge this on the upcoming weekend (after applying the remaining suggestions if they haven't been applied yet).

felixfontein commented:
@jakob-bebop @utoddl thanks for your contributions!
@russoz @jpmens thanks for reviewing!
